pennylane/ops/qubit/hamiltonian.py

```diff
@@ -190,6 +190,15 @@ def __init__(
                 "number of coefficients and operators does not match."
             )
 
+        observables = [
+            (
+                qml.operation.Tensor(*obs.simplify())
+                if isinstance(obs.simplify(), (qml.ops.op_math.Prod))
+                else obs.simplify()
+            )
+            for obs in observables
+        ]
+
```

Comment:
Is this to make legacy Hamiltonian play nice with `Prod`s?
It could be significant performance regression if that is done on every initialization of a legacy Hamiltonian instance 😬 
Do you _need_ this or is this nice to have?

Reply:
> Is this to make legacy Hamiltonian play nice with Prods?
Yep! Ideally, I would like to keep legacy Hamiltonian compatible.

> It could be a significant performance regression
Do you know of any alternative way? I am thinking about it, too.  

> ideally they should never see each other
I think I left a comment about how `hamiltonian` method of `PauliSentence` was still returning `Hamiltonian`, the comparisons being done in those cases with `LinearCombination` were failing.

Reply:
> I think I left a comment about how hamiltonian method of PauliSentence was still returning Hamiltonian, the comparisons being done in those cases with LinearCombination were failing.

I updated that :)

Reply:
see https://github.com/PennyLaneAI/pennylane/pull/5335/commits/7ecea5ac7ad58c33c61c11a139015627a543f1b8

Reply:
When you pull from `ham-tests` do you still get those errors?

Reply:
Looks like I was able to take care of it without the changes!

Reply:
nice :)